# A study of the reliability of borrowers.

### Project Description
The customer is the credit department of a bank. We need to find out whether marital status and the number of children of the client on the fact of repayment of the loan on time. Input data from the bank - statistics on
solvency of clients.

The results of the research will be taken into account when building a model of credit scoring - a special system that evaluates the potential borrower's ability to repay the loan to the bank.


The customer is the credit department of a bank.

It is necessary to find out:
- whether marital status affects the fact of loan repayment on time;
- whether the number of children of the client affects the fact of loan repayment in time.

Input data from the bank - statistics on the solvency of clients.

The results of the research will be taken into account when building a credit scoring model - a special system that assesses the ability of a potential borrower to repay the loan to the bank.

### Step 1: Review the data

In [1]:
#pip install nb_black

In [2]:
%load_ext nb_black

<IPython.core.display.Javascript object>

In [3]:
import pandas as pd

<IPython.core.display.Javascript object>

In [4]:
df = pd.read_csv("/datasets/data.csv")

<IPython.core.display.Javascript object>

In [5]:
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

<IPython.core.display.Javascript object>

In [6]:
df.shape

(21525, 12)

<IPython.core.display.Javascript object>

Let's get general information about df.

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<IPython.core.display.Javascript object>

We see that the dataframe contains data of integer, real and string types. It should also be noted that the columns `days_employed` and `total_income` contain omissions.

### Step 2.1 Filling in the passes

Determine the type of missing values.

In [8]:
df[df["total_income"].isna()].head()

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

<IPython.core.display.Javascript object>

We get that the `days_employed` and `total_income` columns contain missing data of type `NaN`.

Find the number of gaps in these columns.

In [9]:
df["days_employed"].isna().sum()

2174

<IPython.core.display.Javascript object>

In [10]:
df["total_income"].isna().sum()

2174

<IPython.core.display.Javascript object>

Calculate the fraction of missing values in each of the columns with missing values.

In [11]:
df["days_employed"].isna().mean()

0.10099883855981417

<IPython.core.display.Javascript object>

In [12]:
df["total_income"].isna().mean()

0.10099883855981417

<IPython.core.display.Javascript object>

We obtained that the `days_employed` and `total_income` columns have the same share of missing values - 10%. This is due to the fact that monthly income directly depends on the total length of employment. Therefore, if no values are specified in the `days_employed` column, the `total_income` column will also have undefined values.

Оbtain the distribution statistics of the `days_employed` and `total_income` columns.

In [13]:
df["days_employed"].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

<IPython.core.display.Javascript object>

In [14]:
df["total_income"].describe()

count    1.935100e+04
mean     1.674223e+05
std      1.029716e+05
min      2.066726e+04
25%      1.030532e+05
50%      1.450179e+05
75%      2.034351e+05
max      2.265604e+06
Name: total_income, dtype: float64

<IPython.core.display.Javascript object>

We see that in the `days_employed` column the minimum value is negative, which contradicts the name of this column. 
Most likely, an error occurred when filling the dataframe, and some data in the `days_employed` column were added with a minus sign. minus sign. Therefore, let's replace negative values in this column with absolute values.

In [15]:
df["days_employed"] = df["days_employed"].abs()

<IPython.core.display.Javascript object>

In [16]:
df["days_employed"].describe()

count     19351.000000
mean      66914.728907
std      139030.880527
min          24.141633
25%         927.009265
50%        2194.220567
75%        5537.882441
max      401755.400475
Name: days_employed, dtype: float64

<IPython.core.display.Javascript object>

We replace the omissions in these columns with median values because `days_employed` and `total_income` describe quantitative data and some values in these columns stand out strongly among the majority.

Find the median values:

In [17]:
days_employed_median = df["days_employed"].median()

<IPython.core.display.Javascript object>

In [18]:
days_employed_median

2194.220566878695

<IPython.core.display.Javascript object>

In [19]:
total_income_median = df["total_income"].median()

<IPython.core.display.Javascript object>

In [20]:
total_income_median

145017.93753253992

<IPython.core.display.Javascript object>

Print the number of rows with median values in the considered columns.

In [21]:
df[df["days_employed"] == days_employed_median]["days_employed"].count()

1

<IPython.core.display.Javascript object>

In [22]:
df[df["total_income"] == total_income_median]["total_income"].count()

1

<IPython.core.display.Javascript object>

Replace the omissions in the `days_employed` and `total_income` columns with median values.

In [23]:
df.loc[df["days_employed"].isna(), "days_employed"] = days_employed_median

<IPython.core.display.Javascript object>

In [24]:
df.loc[df["total_income"].isna(), "total_income"] = total_income_median

<IPython.core.display.Javascript object>

Count the number of rows with missing values.

In [25]:
df["days_employed"].isna().sum()

0

<IPython.core.display.Javascript object>

In [26]:
df["total_income"].isna().sum()

0

<IPython.core.display.Javascript object>

Print the number of rows with median values in the considered columns.

In [27]:
df[df["days_employed"] == days_employed_median]["days_employed"].count()

2175

<IPython.core.display.Javascript object>

In [28]:
df[df["total_income"] == total_income_median]["total_income"].count()

2175

<IPython.core.display.Javascript object>

Thus, the replacement of missing values with median values was successful, since the final table has 2175 rows with values in columns equal to medians, which is equal to the sum of the number of missing values (2174) and the number of values equal to the median (1) in the original table.

### Step 2.2 Check data for anomalies and corrections.

Let's look at the table again:

In [29]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

<IPython.core.display.Javascript object>

From the definitions of the names of the column names with numeric data of this table it follows that they cannot have negative values, let's check this.

In [30]:
df.select_dtypes(exclude="object")[df.select_dtypes(exclude="object") < 0].count()

children            47
days_employed        0
dob_years            0
education_id         0
family_status_id     0
debt                 0
total_income         0
dtype: int64

<IPython.core.display.Javascript object>

We obtained that the column showing the number of children in the family has negative values. Such values are abnormal.

Find the exact magnitudes of these negative values in the `children` column and their number.

In [31]:
df["children"].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

<IPython.core.display.Javascript object>

We obtained that there is only one negative value -1 in this column. Let's replace it with an absolute value.

In [32]:
df["children"] = df["children"].abs()

<IPython.core.display.Javascript object>

In [33]:
df["children"].value_counts()

0     14149
1      4865
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64

<IPython.core.display.Javascript object>

We now obtain that all values in the `children` column are non-negative. There are no more anomalies in this column.

Take a look at the resulting table.

In [34]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

<IPython.core.display.Javascript object>

It is worth noting that the `dob_years` column cannot have zero values, as the customer must exist to be able to borrow.

In [35]:
df[df["dob_years"] == 0]["dob_years"].count()

101

<IPython.core.display.Javascript object>

Determine the proportion of zero values in this column.

In [36]:
df.loc[df["dob_years"] == 0]["dob_years"].count() / len(df["dob_years"])

0.004692218350754936

<IPython.core.display.Javascript object>

We obtained that the proportion of such values is extremely small, so we can delete the corresponding rows.

In [37]:
df = df.loc[df["dob_years"] != 0]

<IPython.core.display.Javascript object>

In [38]:
df[df["dob_years"] == 0]["dob_years"].count()

0

<IPython.core.display.Javascript object>

It is worth noting that there are only two possible values in the `gender` column: F and M. Let's check if this condition is fulfilled.

In [39]:
df["gender"].value_counts()

F      14164
M       7259
XNA        1
Name: gender, dtype: int64

<IPython.core.display.Javascript object>

We see that there is one line with undefined gender in the dataframe, we can neglect it.

In [40]:
df = df[df["gender"] != "XNA"]

<IPython.core.display.Javascript object>

In [41]:
df["gender"].value_counts()

F    14164
M     7259
Name: gender, dtype: int64

<IPython.core.display.Javascript object>

### Step 2.3. Change the data types.

Replace the real data type in the `total_income` column with an integer one.

In [42]:
df["total_income"] = df["total_income"].astype("int")

<IPython.core.display.Javascript object>

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21423 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21423 non-null  int64  
 1   days_employed     21423 non-null  float64
 2   dob_years         21423 non-null  int64  
 3   education         21423 non-null  object 
 4   education_id      21423 non-null  int64  
 5   family_status     21423 non-null  object 
 6   family_status_id  21423 non-null  int64  
 7   gender            21423 non-null  object 
 8   income_type       21423 non-null  object 
 9   debt              21423 non-null  int64  
 10  total_income      21423 non-null  int64  
 11  purpose           21423 non-null  object 
dtypes: float64(1), int64(6), object(5)
memory usage: 2.1+ MB


<IPython.core.display.Javascript object>

### Step 2.4. Deleting duplicates.

Calculate the number of duplicate rows.

In [44]:
df.duplicated().sum()

54

<IPython.core.display.Javascript object>

Remove the duplicates.

In [45]:
df = df.drop_duplicates().reset_index(drop=True)

<IPython.core.display.Javascript object>

In [46]:
df.duplicated().sum()

0

<IPython.core.display.Javascript object>

Let's look at the table again

In [47]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

<IPython.core.display.Javascript object>

We see that in the `education` column there are values in upper case, which appeared, probably, due to human factor, so let's convert all symbolic values to lower case.

In [48]:
df1 = df.select_dtypes(include="object")

<IPython.core.display.Javascript object>

In [49]:
df1

education     family_status gender income_type  \
0        высшее   женат / замужем      F   сотрудник   
1       среднее   женат / замужем      F   сотрудник   
2       Среднее   женат / замужем      M   сотрудник   
3       среднее   женат / замужем      M   сотрудник   
4       среднее  гражданский брак      F   пенсионер   
...         ...               ...    ...         ...   
21364   среднее  гражданский брак      F   компаньон   
21365   среднее   женат / замужем      F   пенсионер   
21366   среднее  гражданский брак      M   сотрудник   
21367   среднее   женат / замужем      M   сотрудник   
21368   среднее   женат / замужем      F   сотрудник   

                            purpose  
0                     покупка жилья  
1           приобретение автомобиля  
2                     покупка жилья  
3        дополнительное образование  
4                   сыграть свадьбу  
...                             ...  
21364             операции с жильем  
21365          сделка с автомобилем  
21366                  недвижимость  
21367  на покупку своего автомобиля  
21368         на покупку автомобиля  

[21369 rows x 5 columns]

<IPython.core.display.Javascript object>

In [50]:
df1 = df1.apply(lambda x: x.astype(str).str.lower())

<IPython.core.display.Javascript object>

In [51]:
df1

education     family_status gender income_type  \
0        высшее   женат / замужем      f   сотрудник   
1       среднее   женат / замужем      f   сотрудник   
2       среднее   женат / замужем      m   сотрудник   
3       среднее   женат / замужем      m   сотрудник   
4       среднее  гражданский брак      f   пенсионер   
...         ...               ...    ...         ...   
21364   среднее  гражданский брак      f   компаньон   
21365   среднее   женат / замужем      f   пенсионер   
21366   среднее  гражданский брак      m   сотрудник   
21367   среднее   женат / замужем      m   сотрудник   
21368   среднее   женат / замужем      f   сотрудник   

                            purpose  
0                     покупка жилья  
1           приобретение автомобиля  
2                     покупка жилья  
3        дополнительное образование  
4                   сыграть свадьбу  
...                             ...  
21364             операции с жильем  
21365          сделка с автомобилем  
21366                  недвижимость  
21367  на покупку своего автомобиля  
21368         на покупку автомобиля  

[21369 rows x 5 columns]

<IPython.core.display.Javascript object>

In [52]:
df.loc[:, df.select_dtypes(include="object").columns] = df1[df1.columns]

<IPython.core.display.Javascript object>

In [53]:
df

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21364         1    4529.316663         43   среднее             1   
21365         0  343937.404131         67   среднее             1   
21366         1    2113.346888         38   среднее             1   
21367         3    3112.481705         38   среднее             1   
21368         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      f   сотрудник     0   
1       женат / замужем                 0      f   сотрудник     0   
2       женат / замужем                 0      m   сотрудник     0   
3       женат / замужем                 0      m   сотрудник     0   
4      гражданский брак                 1      f   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21364  гражданский брак                 1      f   компаньон     0   
21365   женат / замужем                 0      f   пенсионер     0   
21366  гражданский брак                 1      m   сотрудник     1   
21367   женат / замужем                 0      m   сотрудник     1   
21368   женат / замужем                 0      f   сотрудник     0   

       total_income                       purpose  
0            253875                 покупка жилья  
1            112080       приобретение автомобиля  
2            145885                 покупка жилья  
3            267628    дополнительное образование  
4            158616               сыграть свадьбу  
...             ...                           ...  
21364        224791             операции с жильем  
21365        155999          сделка с автомобилем  
21366         89672                  недвижимость  
21367        244093  на покупку своего автомобиля  
21368         82047         на покупку автомобиля  

[21369 rows x 12 columns]

<IPython.core.display.Javascript object>

In [54]:
df.select_dtypes(include="object").apply(lambda x: x.astype(str).str.isupper()).sum()

education        0
family_status    0
gender           0
income_type      0
purpose          0
dtype: int64

<IPython.core.display.Javascript object>

Thus, all character data in the table is converted to lower case, let's now check for implicit duplicates.

In [55]:
df.duplicated().sum()

17

<IPython.core.display.Javascript object>

Remove the implicit duplicates.

In [56]:
df = df.drop_duplicates().reset_index(drop=True)

<IPython.core.display.Javascript object>

In [57]:
df.duplicated().sum()

0

<IPython.core.display.Javascript object>

In [58]:
df

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21347         1    4529.316663         43   среднее             1   
21348         0  343937.404131         67   среднее             1   
21349         1    2113.346888         38   среднее             1   
21350         3    3112.481705         38   среднее             1   
21351         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      f   сотрудник     0   
1       женат / замужем                 0      f   сотрудник     0   
2       женат / замужем                 0      m   сотрудник     0   
3       женат / замужем                 0      m   сотрудник     0   
4      гражданский брак                 1      f   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21347  гражданский брак                 1      f   компаньон     0   
21348   женат / замужем                 0      f   пенсионер     0   
21349  гражданский брак                 1      m   сотрудник     1   
21350   женат / замужем                 0      m   сотрудник     1   
21351   женат / замужем                 0      f   сотрудник     0   

       total_income                       purpose  
0            253875                 покупка жилья  
1            112080       приобретение автомобиля  
2            145885                 покупка жилья  
3            267628    дополнительное образование  
4            158616               сыграть свадьбу  
...             ...                           ...  
21347        224791             операции с жильем  
21348        155999          сделка с автомобилем  
21349         89672                  недвижимость  
21350        244093  на покупку своего автомобиля  
21351         82047         на покупку автомобиля  

[21352 rows x 12 columns]

<IPython.core.display.Javascript object>

Check the number of unique values in the `education` and `education_id` columns.

In [59]:
df["education"].value_counts()

среднее                15108
высшее                  5215
неоконченное высшее      741
начальное                282
ученая степень             6
Name: education, dtype: int64

<IPython.core.display.Javascript object>

In [60]:
df["education_id"].value_counts()

1    15108
0     5215
2      741
3      282
4        6
Name: education_id, dtype: int64

<IPython.core.display.Javascript object>

It can be seen that the number of identifiers and the number of values describing the level of education are the same.

Check the number of unique values in the `family_status` and `family_status_id` columns.

In [61]:
df["family_status"].value_counts()

женат / замужем          12290
гражданский брак          4129
не женат / не замужем     2794
в разводе                 1185
вдовец / вдова             954
Name: family_status, dtype: int64

<IPython.core.display.Javascript object>

In [62]:
df["family_status_id"].value_counts()

0    12290
1     4129
4     2794
3     1185
2      954
Name: family_status_id, dtype: int64

<IPython.core.display.Javascript object>

Again we got the same number of identifiers and values.

### Step 2.5. Formation of additional dictionary dataframes, decomposition of the initial dataframe.

Let's create two new dataframes "dictionaries".

In [63]:
df_1 = df[["education_id", "education"]]
df_1

education_id education
0                 0    высшее
1                 1   среднее
2                 1   среднее
3                 1   среднее
4                 1   среднее
...             ...       ...
21347             1   среднее
21348             1   среднее
21349             1   среднее
21350             1   среднее
21351             1   среднее

[21352 rows x 2 columns]

<IPython.core.display.Javascript object>

Remove duplicates in df_1.

In [64]:
df_1.duplicated().sum()

21347

<IPython.core.display.Javascript object>

In [65]:
df_1 = df_1.drop_duplicates().reset_index(drop=True)
df_1

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

<IPython.core.display.Javascript object>

In [66]:
df_2 = df[["family_status_id", "family_status"]]
df_2

family_status_id     family_status
0                     0   женат / замужем
1                     0   женат / замужем
2                     0   женат / замужем
3                     0   женат / замужем
4                     1  гражданский брак
...                 ...               ...
21347                 1  гражданский брак
21348                 0   женат / замужем
21349                 1  гражданский брак
21350                 0   женат / замужем
21351                 0   женат / замужем

[21352 rows x 2 columns]

<IPython.core.display.Javascript object>

Remove duplicates in df_2.

In [67]:
df_2.duplicated().sum()

21347

<IPython.core.display.Javascript object>

In [68]:
df_2 = df_2.drop_duplicates().reset_index(drop=True)
df_2

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

<IPython.core.display.Javascript object>

Remove the `education` and `family_status` columns from the original dataframe.

In [69]:
df.drop(["education", "family_status"], axis=1, inplace=True)
df

children  days_employed  dob_years  education_id  family_status_id  \
0             1    8437.673028         42             0                 0   
1             1    4024.803754         36             1                 0   
2             0    5623.422610         33             1                 0   
3             3    4124.747207         32             1                 0   
4             0  340266.072047         53             1                 1   
...         ...            ...        ...           ...               ...   
21347         1    4529.316663         43             1                 1   
21348         0  343937.404131         67             1                 0   
21349         1    2113.346888         38             1                 1   
21350         3    3112.481705         38             1                 0   
21351         2    1984.507589         40             1                 0   

      gender income_type  debt  total_income                       purpose  
0          f   сотрудник     0        253875                 покупка жилья  
1          f   сотрудник     0        112080       приобретение автомобиля  
2          m   сотрудник     0        145885                 покупка жилья  
3          m   сотрудник     0        267628    дополнительное образование  
4          f   пенсионер     0        158616               сыграть свадьбу  
...      ...         ...   ...           ...                           ...  
21347      f   компаньон     0        224791             операции с жильем  
21348      f   пенсионер     0        155999          сделка с автомобилем  
21349      m   сотрудник     1         89672                  недвижимость  
21350      m   сотрудник     1        244093  на покупку своего автомобиля  
21351      f   сотрудник     0         82047         на покупку автомобиля  

[21352 rows x 10 columns]

<IPython.core.display.Javascript object>

### Step 2.6. Categorization of income.

Based on the ranges below, let's create a column
`total_income_category` with categories:
- 0-30000 - 'E';
- 30001-50000 - 'D';
- 50001-200000 - 'C';
- 200001-1000000 - 'B';
- 1000001 and above - 'A'.

For example, a borrower with an income of 25000 should be
would be categorized as 'E' and a client earning 235,000, -'B'.

In [70]:
def group_income(income):
    if income <= 30000:
        return "E"
    elif income >= 30001 and income <= 50000:
        return "D"
    elif income >= 50001 and income <= 200000:
        return "C"
    elif income >= 200001 and income <= 1000000:
        return "B"
    else:
        return "A"

<IPython.core.display.Javascript object>

In [71]:
df["total_income_category"] = df["total_income"].apply(group_income)

<IPython.core.display.Javascript object>

In [72]:
df

children  days_employed  dob_years  education_id  family_status_id  \
0             1    8437.673028         42             0                 0   
1             1    4024.803754         36             1                 0   
2             0    5623.422610         33             1                 0   
3             3    4124.747207         32             1                 0   
4             0  340266.072047         53             1                 1   
...         ...            ...        ...           ...               ...   
21347         1    4529.316663         43             1                 1   
21348         0  343937.404131         67             1                 0   
21349         1    2113.346888         38             1                 1   
21350         3    3112.481705         38             1                 0   
21351         2    1984.507589         40             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          f   сотрудник     0        253875                 покупка жилья   
1          f   сотрудник     0        112080       приобретение автомобиля   
2          m   сотрудник     0        145885                 покупка жилья   
3          m   сотрудник     0        267628    дополнительное образование   
4          f   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21347      f   компаньон     0        224791             операции с жильем   
21348      f   пенсионер     0        155999          сделка с автомобилем   
21349      m   сотрудник     1         89672                  недвижимость   
21350      m   сотрудник     1        244093  на покупку своего автомобиля   
21351      f   сотрудник     0         82047         на покупку автомобиля   

      total_income_category  
0                         B  
1                         C  
2                         C  
3                         B  
4                         C  
...                     ...  
21347                     B  
21348                     C  
21349                     C  
21350                     B  
21351                     C  

[21352 rows x 11 columns]

<IPython.core.display.Javascript object>

In [73]:
df["total_income_category"].value_counts(dropna=False)

C    15943
B     5014
D      348
A       25
E       22
Name: total_income_category, dtype: int64

<IPython.core.display.Javascript object>

In [74]:
df["total_income_category"].count()

21352

<IPython.core.display.Javascript object>

We see that the function worked correctly. Now in our dataframe each client of the bank belongs to a certain group depending on income.

### Step 2.7. Categorization of credit objectives.

Let's create a function that will form a new column `purpose_category` based on the data from the column purpose will form a new column `purpose_category`, which will include the following categories:
- 'automobile transactions',
- 'real estate transactions',
- 'wedding',
- 'education'.

Let's look at the different values of the `purpose` column.

In [75]:
df["purpose"].value_counts()

свадьба                                   786
на проведение свадьбы                     764
сыграть свадьбу                           760
операции с недвижимостью                  672
покупка коммерческой недвижимости         658
покупка жилья для сдачи                   649
операции с коммерческой недвижимостью     648
операции с жильем                         646
жилье                                     640
покупка жилья                             640
покупка жилья для семьи                   637
строительство собственной недвижимости    633
недвижимость                              629
операции со своей недвижимостью           627
строительство жилой недвижимости          621
строительство недвижимости                619
покупка своего жилья                      619
покупка недвижимости                      617
ремонт жилью                              605
покупка жилой недвижимости                603
на покупку своего автомобиля              502
заняться высшим образованием      

<IPython.core.display.Javascript object>

In [76]:
def group_purpose(purpose):
    if "авто" in purpose:
        return "операции с автомобилем"
    elif "образ" in purpose:
        return "получение образования"
    elif "свадь" in purpose:
        return "проведение свадьбы"
    else:
        return "операции с недвижимостью"

<IPython.core.display.Javascript object>

In [77]:
df["purpose_category"] = df["purpose"].apply(group_purpose)

<IPython.core.display.Javascript object>

In [78]:
df

children  days_employed  dob_years  education_id  family_status_id  \
0             1    8437.673028         42             0                 0   
1             1    4024.803754         36             1                 0   
2             0    5623.422610         33             1                 0   
3             3    4124.747207         32             1                 0   
4             0  340266.072047         53             1                 1   
...         ...            ...        ...           ...               ...   
21347         1    4529.316663         43             1                 1   
21348         0  343937.404131         67             1                 0   
21349         1    2113.346888         38             1                 1   
21350         3    3112.481705         38             1                 0   
21351         2    1984.507589         40             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          f   сотрудник     0        253875                 покупка жилья   
1          f   сотрудник     0        112080       приобретение автомобиля   
2          m   сотрудник     0        145885                 покупка жилья   
3          m   сотрудник     0        267628    дополнительное образование   
4          f   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21347      f   компаньон     0        224791             операции с жильем   
21348      f   пенсионер     0        155999          сделка с автомобилем   
21349      m   сотрудник     1         89672                  недвижимость   
21350      m   сотрудник     1        244093  на покупку своего автомобиля   
21351      f   сотрудник     0         82047         на покупку автомобиля   

      total_income_category          purpose_category  
0                         B  операции с недвижимостью  
1                         C    операции с автомобилем  
2                         C  операции с недвижимостью  
3                         B     получение образования  
4                         C        проведение свадьбы  
...                     ...                       ...  
21347                     B  операции с недвижимостью  
21348                     C    операции с автомобилем  
21349                     C  операции с недвижимостью  
21350                     B    операции с автомобилем  
21351                     C    операции с автомобилем  

[21352 rows x 12 columns]

<IPython.core.display.Javascript object>

In [79]:
df["purpose_category"].value_counts(dropna=False)

операции с недвижимостью    10763
операции с автомобилем       4284
получение образования        3995
проведение свадьбы           2310
Name: purpose_category, dtype: int64

<IPython.core.display.Javascript object>

In [80]:
df["purpose_category"].count()

21352

<IPython.core.display.Javascript object>

We see that the function worked correctly. Now in our dataframe each client of the bank belongs to a certain group depending on the purpose of the loan.

### Answering valuable questions.

##### Is there a correlation between the number of children and paying back the loan on time?

In [81]:
df["children"].unique()

array([ 1,  0,  3,  2,  4, 20,  5])

<IPython.core.display.Javascript object>

In [82]:
pd.pivot_table(df, index="children", values="debt", aggfunc=["count", "mean"])

count      mean
           debt      debt
children                 
0         14021  0.075458
1          4839  0.091341
2          2039  0.095145
3           328  0.082317
4            41  0.097561
5             9  0.000000
20           75  0.106667

<IPython.core.display.Javascript object>

Break down the data so that there are a comparable number of people in each age group.

In [83]:
df["children_group1"] = pd.cut(df["children"], [-float("inf"), 0.9, 1.9, float("inf")])

<IPython.core.display.Javascript object>

In [84]:
df.groupby("children_group1")["debt"].agg(["count", "mean"])

count      mean
children_group1                 
(-inf, 0.9]      14021  0.075458
(0.9, 1.9]        4839  0.091341
(1.9, inf]        2492  0.093499

<IPython.core.display.Javascript object>

##### Answer 1:

There is a correlation between the number of children and loan repayment on time. The more children, the higher the risk that the loan will not be repaid on time. The most reliable borrowers will be people who have no children, the least reliable - those who have more than 1 child.

##### Is there a correlation between marital status and loan repayment on time?

There are 5 types of marital status in total, which are summarized in the "dictionary" table df_2:

In [85]:
df_2

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

<IPython.core.display.Javascript object>

In [86]:
df.groupby("family_status_id")["debt"].agg(["count", "mean"]).sort_values(
    "mean", ascending=False
)

count      mean
family_status_id                 
4                  2794  0.097709
1                  4129  0.093485
0                 12290  0.075427
3                  1185  0.071730
2                   954  0.064990

<IPython.core.display.Javascript object>

##### Answer 2:

We get that people who were not married were the most likely to default on their payments, and widows and widowers were the least likely to default.

##### Is there a correlation between income level and loan repayment on time?

In [87]:
df.groupby("total_income_category")["debt"].agg(["count", "mean"])

count      mean
total_income_category                 
A                         25  0.080000
B                       5014  0.070802
C                      15943  0.084865
D                        348  0.060345
E                         22  0.090909

<IPython.core.display.Javascript object>

##### Answer 3:

This table shows that the probability of loan repayment at maturity does not depend on the borrower's income level. The amount of data in categories B and C is sufficient to conclude that the most reliable borrowers are from group B, and the least reliable are from group C.

##### How do the different purposes of a loan affect its repayment on time?

In [88]:
df.groupby("purpose_category")["debt"].agg(["count", "mean"])

count      mean
purpose_category                         
операции с автомобилем     4284  0.093371
операции с недвижимостью  10763  0.072378
получение образования      3995  0.092616
проведение свадьбы         2310  0.079654

<IPython.core.display.Javascript object>

##### Answer 4:

We see that the probability of repaying a loan taken for real estate transactions and weddings at maturity is on average higher than in the case of car transactions and education.

## Overall conclusion:

As a result of the conducted research, the dependence of the fact of loan repayment on time on the following parameters was obtained:
1. Marital status,
2. Number of children of the client,
3. The purpose of the loan.

There is no direct correlation between the level of income and the probability of repayment of the loan on time.

This result was obtained from a preprocessed dataframe. As a result of data preprocessing, negative values in the `children` and `days_employed` columns were replaced with absolute values, gaps in the `days_employed` and `total_income` columns were filled with median values, data on customers with zero age were removed, and only two types of values were left in the `gender` column: M and F. Also, explicit duplicates were removed, and then implicit duplicates were removed after the string data were lower-cased.